homeassistant/components/humidifier/__init__.py

```diff
@@ -220,7 +220,7 @@ def action(self) -> HumidifierAction | None:
         return self._attr_action
 
     @cached_property
-    def current_humidity(self) -> int | None:
+    def current_humidity(self) -> float | None:
```

Comment:
The current humidity is documented as an optional integer.

https://developers.home-assistant.io/docs/core/entity/humidifier#properties

The issue says the sensor is accurate to +/- 1 %. Does that mean it isn't actually more accurate than an integer would represent?

Would a home user realistically be interested in and have a sensor that can meaure less than a percent of accuracy of humidity? That sounds far fetched.

Reply:
> The issue says the sensor is accurate to +/- 1 %. Does that mean it isn't actually more accurate than an integer would represent?
> 
> Would a home user realistically be interested in and have a sensor that can meaure less than a percent of accuracy of humidity? That sounds far fetched.

Not quite. Even if a sensor spec is +/- 1% that doesn't necessarily mean its accuracy isn't higher. Might just have a built in tolerance of up to 1% in each direction.

The main thing to consider though is that we don't do any kind of smoothing when converting the values from float to int. That's why the graph looks a lot more fragmented after the update. Of course it would be possible to apply that afterwards but due to the initial loose of accuracy, you would now be looking more at something like +/- 2%.

/CC @kkalitowski

> The current humidity is documented as an optional integer.
>
> https://developers.home-assistant.io/docs/core/entity/humidifier#properties

I can update that once we decide this change is fine.

Reply:
I still don't see the need or interest for higher accuracy. It sounds like it's just a problem with how the graph looks, not an actual desire to measure more accurate.

Reply:
I appreciate the dialog and constructive thoughts. Here is why I think a higher resolution than int would be beneficial:

1. With integer resolution, the chart is difficult to read and trends are hard to see. If the sensor reads 32.48, 32.52, 32.48, the chart bounces from 32-33-32.
2. I have the wet_tolerance on the humidifier set to 1 because I do not want to overdo humidity and have it too high when night comes (it takes a long time to come down). With integer resolution, these tolerances take on a different meaning. A tolerance of 1% now becomes 0.5% due to rounding.
3. Take items 1 and 2 together, and you find that the humdifier might be going on/off every minute, putting wear on the physical components (relay, solenoid, jiggling pipes) and causing extra noise in the house from the pipes. I could compensate with generic_hygrostat's min_cycle_duration, but that strikes me as a workaround, not the real answer to the problem.

Reply:
You can increase the tolerance to 2 % and avoid the problem. Since your sensor isn't more accurate than 1 % it sounds appropriate to set a higher tolerance than 1.

Reply:
The [Sensiron SHT45](https://sensirion.com/media/documents/33FD6951/6555C40E/Sensirion_Datasheet_SHT4x.pdf) sensor has an accuracy of +/- 1.0% RH, a resolution of 0.01% RH, and repeatability of 0.08% RH.

Yes, it _could_ be off by 1% in either direction depending on the environment or manufacturing issues, but it is mostly consistent and precise in the measurement it provides. It won't return three readings of 30, 29, 31; it will give three readings within 0.08% of each other.

Personally, I think it should be up to the user to decide how precise of a value to send to `current_humidity` instead of being artificially limited to `int`. For one, it makes the chart much more readable so I can have one humidifier chart on the dashboard instead of needing a second chart to also see the more precise data.

But beyond the chart, the rounding adversely affects the functionality and understandability of the humidifier.  With a `dry_tolerance` and `wet_tolerance` each of 1, you are only getting a _total_ range of 1. That’s confusing and narrow. If you bump tolerance to 2, you are getting a range of 3. On the other hand, the cases that use `float` make natural sense and have no chance of causing confusion. Anyone with a sensor resolution better than an integer, which I expect is at least everyone using ESPHome, will have this problem.

Target RH% | Numeric | Wet/dry Tolerance | Real range | Humidifier turns on | Humidifier turns off
---------- | ------------ | --------- | ---------- | ------------------- | --------------------
30         | int          | 1 / 1         | 1.01       | 29.49               | 30.5                
30         | int          | 2 / 2         | 3.01       | 28.49               | 31.5                
30         | float        | 1 / 1         | 2          | 29                  | 31                  
30         | float        | 2 / 2         | 4          | 28                  | 32                  

<I retract my item 3 from my earlier reply; I don't think using `int` is going to cause major problems with a humidifier cycling, though it will still increase cycling because of the rounding>

Reply:
We've discussed this in the core team and concluded it's ok to change the type to float.

Side note: We think there should be a way for humidifer entities to say what the default precision shown to the user is, similar to how sensor entities have a `suggested_precision` attribute.